e2e_test/udf/sql_udf.slt

```diff
@@ -72,6 +80,19 @@ select call_regexp_replace();
 ----
 💩💩💩💩💩foo🤔️bar亲爱的😭这是🥵爱情❤️‍🔥
 
+query T
+select foo(114514);
+----
+foo(INT)
+
+# Rejected deep calling stack
+statement error function recursive calling stack depth limit exceeded
+select recursive(1, 1);
+
+# Same as above
+statement error function recursive calling stack depth limit exceeded
+select recursive_non_recursive(1, 1);
```

Comment:
Can you add
- a case for create MV
- a case for a valid (terminating) recursive function e.g., 

```sql
create function fib(bigint) returns bigint
  as 'select case
        when $1 = 0 then 0
        when $1 = 1 then 1
        when $1 = 2 then 1
        when $1 = 3 then 2
      else fib($1-1) + fib($1-2)
      end;'
language sql
```

Reply:
Regarding the `fib` example above, I think the error is not very user friendly. 🤣

One thing is: "function calling stack depth" is a little confusing for this example.

```
dev=> select fib(1::bigint);
ERROR:  Failed to run the query

Caused by these errors (recent errors listed first):
  1: Failed to bind expression: fib(CAST(1 AS BIGINT))
  2: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
  3: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
  4: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
  5: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
  6: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
  7: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
  8: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
  9: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 10: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 11: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 12: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 13: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 14: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 15: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 16: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 17: Failed to bind expression: CASE WHEN $1 = 0 THEN 0 WHEN $1 = 1 THEN 1 WHEN $1 = 2 THEN 1 WHEN $1 = 3 THEN 2 ELSE fib($1 - 1) + fib($1 - 2) END
 18: Bind error: function fib calling stack depth limit exceeded
```

Reply:
Well, I think it's still acceptable.

Reply:
Quite similar to the original Postgres experience! 😄

```
postgres=# select recursive(114, 514);
ERROR:  stack depth limit exceeded
HINT:  Increase the configuration parameter "max_stack_depth" (currently 2048kB), after ensuring the platform's stack depth limit is adequate.
CONTEXT:  SQL function "recursive" during inlining
SQL function "recursive" during startup
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
SQL function "recursive" statement 1
...
```

Reply:
> a case for a valid (terminating) recursive function e.g.,

Tried to implement this but eventually found that it's sort of impossible to handle this under current binder framework (unless we evaluate the expression during binding). The reason is that, we are **inlining** the corresponding logic rather than **evaluating** it in the binder, which will then lead to, for example, in the case-when example, infinite binding in `bind_case` / `bind_function` / ..., the essential reason for this is that, in `bind_case`, we can not only evaluating the base case and **directly** return, and the binding in `bind_case` later, i.e.,
```rust
let else_result_expr = else_result
            .map(|expr| self.bind_expr_inner(*expr))
            .transpose()?;
```
will inevitably cause the infinite recursive binding, even though we meet the base case.

Reply:
> a case for create MV

I don't get it, could you elaborate on this?

Reply:
I mean create a MV with a recursive function.

Reply:
I'm not asking to implement these features, but want these to be covered in test cases so that we know the behavior.

Reply:
> The reason is that, we are **inlining** the corresponding logic rather than **evaluating** it in the binder

Does this mean that we're not able to support any kind of recursion in SQL UDF? If so, what about directly rejecting recursion?

Reply:
I also raised this point before. 

https://github.com/risingwavelabs/risingwave/pull/14139#discussion_r1439219709

I thought it's easy, but it turns out at least not trivial. 🤔

Reply:
If that's not too hard, I suggest we can go that way to make things easier (possibly).

Reply:
> I thought it's easy, but it turns out at least not trivial. 

The function check, as for pg, will reject calling other non-existence function during definition time, but recursion definition is totally valid, and could be actually run. 🤡

As for rw, I decide to put this feature to later PR, it's not hard to implement, but not quite related to this PR. (i.e., It's preventing recursive calling chain at definition time, but this PR aims to resolve the cluster panic problem that may exist with deep calling stack)

Reply:
> The function check, as for pg, will reject calling other non-existence function during definition time, but recursion definition is totally valid, and could be actually run. 🤡 (P.S. This is why I think we should manage to support recursion for sql udf)

I know. To clarify, I mean we can choose not to support this feature to make our lives easier, just like recursive CTE. 🤣

Reply:
I see, then let's just forbid recursive sql udf during runtime at present, maybe support in the future.🤪